File: csharp/DotnetLambdaBenchmark/src/DotnetLambdaBenchmark/Bucket.cs
using System;
using System.Threading.Tasks;

using Amazon.S3;
using Amazon.S3.Model;


namespace DotnetLambdaBenchmark
{
    class Bucket: IDisposable{
        public Bucket(string bucketName) {
            this.client = new AmazonS3Client();
            this.bucketName = bucketName;
        }

        public async Task<bool> CheckObjectPresence(FilePath path) {
            try {
                await client.GetObjectMetadataAsync(this.bucketName, path.NormalizedPath);
                return true;
            }
            catch (Exception) {
                return false;
            }
        }

        public async Task CreateObject(FilePath path) {
            try {
                var response = await client.PutObjectAsync(new PutObjectRequest(){
                    BucketName = this.bucketName,
                    Key = path.NormalizedPath,
                });
            }
            catch(Exception e) {
                Console.WriteLine($"Failed to put object into S3 with error: {e.Message}");
                throw new ServerException("Failed to create S3 object");
            }
        }

        public void Dispose()
        {
            client.Dispose();
        }

        IAmazonS3 client;
        string bucketName;
    }

}

File: java/src/main/java/benchmark/Bucket.java
package benchmark;

import java.lang.System;
import java.util.concurrent.CompletableFuture;

import software.amazon.awssdk.core.async.AsyncRequestBody;
import software.amazon.awssdk.services.s3.*;
import software.amazon.awssdk.services.s3.model.HeadObjectRequest;
import software.amazon.awssdk.services.s3.model.PutObjectRequest;
import software.amazon.awssdk.services.s3.model.S3Exception;

class Bucket {
    public Bucket(String bucketName) {
        this.client = S3AsyncClient.create();
        this.bucketName = bucketName;
    }

    public CompletableFuture<Boolean> CheckObjectPresence(FilePath path) {
            return this.client.headObject(
                HeadObjectRequest
                    .builder()
                    .bucket(this.bucketName)
                    .key(path.getNormalized())
                    .build()
            )
            .handle((response, error) -> error == null);
    }

    public CompletableFuture<Void> CreateObject(FilePath path) {
        return this.client.putObject(
                PutObjectRequest
                        .builder()
                        .bucket(this.bucketName)
                        .key(path.getNormalized())
                        .build(),
                AsyncRequestBody.empty()
        ).handle((response, error) -> {
            if (error instanceof S3Exception) {
                System.out.printf("Failed to put object into S3 with error: %s", error.toString());
                throw new ServerException("Failed to create S3 object");
            }
            return null;
        });
    }

    S3AsyncClient client;
    String bucketName;
}
